### ISSUE TITLE
"reportImplicitOverride" gets confused by wrapper function with inferred return type

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#5133]
**Describe the bug**
If result type of a wrapper function is omitted (so that a function can be returned instead of a callable), and an overriding method is annotated by it, the returned function inside the wrapper function is incorrectly required to be annotated with `@override`.

**To Reproduce**
See sample code below.

As a non-overriding method and an overriding method get annotated with a same function, adding a `@override` or not will both get an error.

**Expected behavior**
`__eq__` needs `@override`. `wrapped` inside `wrap` doesn't need `@override`.

**Screenshots or Code**
```python
# pyright: strict
# pyright: reportImplicitOverride=true

from __future__ import annotations

from typing import Any, Callable, TypeVar
from typing_extensions import override

T = TypeVar("T")

def wrap(func: Callable[..., Any], /):
    def wrapped(*args: Any, **kwargs: Any) -> Any:
        raise NotImplementedError
    
    return wrapped

class C:
    @wrap
    def f(self, /) -> None:
        raise NotImplementedError
    
    @override
    @wrap
    def __eq__(self, other: object, /) -> bool:
        raise NotImplementedError
```
![image](https://github.com/microsoft/pyright/assets/15389744/404ed0eb-0a5c-4f6d-bf0d-3a26c0fc86cd)
![image](https://github.com/microsoft/pyright/assets/15389744/9ea9417a-2a57-43ee-a853-77b8d1c6f6b8)

**VS Code extension or command-line**
Pyright v1.1.306 ~ v1.1.309

**Additional context**
N/A

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.310, which I just published. It will also be included in a future release of pylance.
